### DISCUSSION TITLE
Ajout de script pour un champ texte

### DISCUSSION THREAD

**Discussion by jodionrq** [MTESSDev/FRW#84] [Annonces]
Bonjour,
Dans l'exemple du P700U, il y a des exemples d'injection de script pour des champs de date. Est-ce possible d'ajouter un script en passant par Method et Computed pour des champs de texte où l'utilisateur doit entrer une valeur?
Dans mon cas, j'ai champ de texte avec une validation en année (yyyy). J'aimerais faire apparaitre un message si l'utilisateur entre une année égale ou -1 an à l'année en cours. Exemple: Nous sommes en 2023, alors si l'utilisateur entre "2023" ou "2022", un message devrait apparaitre.
![image](https://github.com/MTESSDev/FRW/assets/124822406/0731b013-4a1d-4195-a176-d6f6da70900b)

Si ce n'est pas possible, est-ce qu'on peut ajouter une validation avec un v-if ?
Si vous avez des questions, n'hésitez pas à communiquer avec moi.
Merci à l'avance!

**Comment by j0ber** [comment]
Bonjour,
Dans votre exemple, le message est un message d'erreur (i.e. une validation avec message d'erreur en rouge sous le champ et une erreur à la révision du formulaire) ? Ou bien c'est juste un avertissement ou autre?

**Reply by jodionrq** [reply]
Bonjour,
C'est un avertissement à la sortie du champ. Ça ne fait pas un message d'erreur mais ça guide l'utilisateur avant de continuer de remplir le reste du formulaire. Aussi, je dois préciser que le calcul de moins d'un an se fait par rapport au 1er septembre de l'année en cours. Donc j'aurais ces valeurs à calculer : 
1. La date actuelle; 
2. Si l'année saisie est égale à l'année en cours -> Avertissement;
3. Si l'année saisie est égale à -1 an avec l'année en cours :
    - Si la date actuelle est avant le 1er septembre de l'année en cours -> Avertissement.

**Comment by j0ber** [comment]
Dans ce cas il faut effectivement injecter du script. Dans votre cas ce serait une computed (ex. doitAfficherMessageAvertissementDate). Cette dernière va contenir le code qui va comparer la valeur du champ avec la condition que vous voulez et retourner true/false selon que la valeur saisie est valide ou non. 
Bref, vous procédez en injectant une computed comme dans l'exemple. Et sur votre message d'avertissement vous mettez un v-if qui utilise votre computed. Tout le code pourrait être dans le v-if (donc pas de computed) mais ça risque d'être plus qu'une simple petite condition, donc une computed est préférable.

**Reply by jodionrq** [reply]
Merci beaucoup pour la réponse.
Bonne journée!

**Reply by eparentrq** [reply]
Bonjour,
              voici la computed et le v-if. Rien ne fonctionne.


![computed](https://github.com/MTESSDev/FRW/assets/133912609/01fa0388-17a3-4827-baeb-1d1bd4829058)
![v_if](https://github.com/MTESSDev/FRW/assets/133912609/385348c1-7d67-480a-b7ed-ccaa9aa7fd62)

**Comment by j0ber** [comment]
Déjà votre v-if n'est pas correct... c'est du javascript. Là vous faite une assignation et non une comparaison.

Votre v-if devrait être :  
v-if: differenceAnnee === false   mais encore mieux puisque differenceAnnee retourne un booleen -> v-if: !differenceAnnee

Aussi si vous avez consulté les [exemples](https://formulaires.it.mtess.gouv.qc.ca/fr/Form/7/P700U/0/N/#p=sectionInjecterJs), vous verrez que lorsque vous faites référence à une méthode ou une computed à l'intérieur d'une autre méthode ou computed, vous devez préfixer de "this".

Dans votre cas if(calculerDifferenceAnnee(...)     devrait être if(this.calculerDifferenceAnnee(...)

De plus il peut aussi y avoir une erreur dans "calculerDiffenrenceAnnee" mais je n'ai pas le code.

S.V.P. utilisez la fonction "Add code" (ctrl+E) de l'éditeur texte afin de coller votre code. Nous pourrons ainsi le copier/coller et tester.
De plus aussi inclure la définition des champs du formulaire impliqués afin que nous puissions reproduire le cas.

Mais déjà, vous pouvez corriger les 2 problèmes que j'ai mentionné et voir ce que ça donne. 

Merci

**Reply by eparentrq** [reply]
Merci beaucoup :-)

Petite question. Est-ce qu'il y a un moyen d'avoir un debugger?
Merci.

**Comment by eparentrq** [comment]
# Groupes pour chaque année (formulaire)
        - section:
            fr: Correction aux revenus de travail selon l'année
            en: Salary
          id: infosSalarial
          components:
              - type: repeatableGroup
                name: groupeannee
                repeatable: true
                label:
                  fr: Pour faire corriger vos revenus de travail des années passées, vous devez remplir le tableau ci-dessous pour chaque année en cause.
                instanceLabel:
                  fr: Année visée
                addLabel:
                  fr: Ajouter une année
                removeLabel:
                  fr: Supprimer une année
                components:
                  - name: dateAnnee
                    type: text
                    label: 
                      fr: Année
                      en: Year
                    validations:
                      date: YYYY
                    outerClasses: sm
                    







        # Je pense que le calcul résulte une différence d'année et non une différence en chiffre.                   
                  - type: dynamic
                    tag: div
                    classes: mb-32
                    v-if: nbAnneeDifference === 0
                    text:
                      fr: Même année que l'année en cours`



Avec ce code, je ne vois plus mon groupe...

`  computed:
    nbAnneeDifference: 
      code: |
        ()

          return 3;`

**Reply by j0ber** [reply]
j'ai besoin de tout le code et la config impliqués. À la limite votre fichier yaml en PJ encore plus simple

**Reply by eparentrq** [reply]
Malheureusement, les fichiers .yaml ne sont pas acceptés.

**Reply by j0ber** [reply]
Simplement renommer votre fichier yml en txt et le coller

**Reply by eparentrq** [reply]
[RDC025.v15.3.form.txt](https://github.com/MTESSDev/FRW/files/11508995/RDC025.v15.3.form.txt)

voilà!

**Reply by j0ber** [reply]
merci je vais regarder ça quand j'aurai quelques minutes... 
Pour le debuggage, le plus simple est de déployer votre formulaire. Ensuite vous pouvez utiliser les outils de debuggage de n'importe quel fureteur.

**Reply by j0ber** [reply]
Bon... Plusieurs choses...

D'abord c'est du YAML. L'indentation est IMPORTANTE (j'ai aussi de la misère avec ça :)   )

Votre injecterJs ne fonctionne pas c'est certain... Aucune de vos computed ou méthode ne peut fonctionner... InjecterJS doit être dans "Config".

On voyait clairement dans la console que quelque chose clochait avec ça :  
![image](https://github.com/MTESSDev/FRW/assets/26974817/13cf5caa-434f-4e82-9141-4774207d3f82)

Il faut augmenter l'indentation de votre bloc "InjecterJS" pour qui soit à l'intérieur de "Config".

Je constate également que vous avez des groupes dans un groupe... **Ce n'est pas supporté**. Je constate que nous n'avons pas documenté cela, nous allons l'adresser. Vous avez l'impression que ça fonctionne mais quand vous allez valider tout va exploser...

Vous devrez revoir votre découpage.

Un groupe ne devrait pas contenir autant d'information, ça ne fait pas vraiment de sens.

Aussi je constate que vous mettez tous vos composants dans dans composants "dynamic". 

C'est lourd inutilement. Les dynamic devraient être utilisés pour afficher du html ou un groupe de contrôles et appliquer une condition v-if à tous les éléments qu'il contient.

En gros, un "dynamic" qui contient un seul component c'est inutile. La condition v-if devrait être appliquée sur le component directement. Nul besoin de l'encapsuler dans un "dynamic".

Pour conclure, nous aurions des suggestions pour revoir votre découpage afin d'éviter d'avoir un groupe dans un groupe.
Nous pourrions en discuter dans une rencontre de travail.

**Reply by eparentrq** [reply]
Merci beaucoup.

J'ai envoyé votre réponse à un collègue, puisque c'est lui au départ qui avait fait une demande. Je ne suis qu'une aide (option de secours ;))

Je lui laisse le soin de vous contacter ou sinon, je prendrai la relève. Merci pour votre retour rapide.

**Reply by jodionrq** [reply]
Bonjour,
Je soupçonnais que le groupe dans un groupe n'était pas supporté, car le message d'erreur à la validation d'un champ obligatoire vide nous donnais "true" comme message.
Nous allons modifier le formulaire pour retirer ce groupe répétable et vérifier si notre JS fonctionne bien.
Merci pour votre offre de suggestions c'est très gentil, mais nous sommes proche des lab clients et de notre livraison, alors on va procéder avec cette version (sans les groupes dans un groupe) et nous pourrons apporter des améliorations pour une autre version.
Je ferai une autre demande si j'ai d'autres question pour le JS.
Un gros merci pour votre collaboration!

**Reply by eparentrq** [reply]
Bonjour, merci @j0ber énormément pour votre aide.
J'ai une question.
Comment dois-je m'y prendre pour exécuter une vérification après que le champ texte soit modifié?
Le client doit entrer une année dans un champ texte et des vérifications doivent être faite selon ce qu'il a inscrit.

Merci de m'aider.

**Reply by j0ber** [reply]
Il n'y a rien de particulier à faire, la vérification va s'exécuter toute seule. Ex. Vous avez un dynamic avec un v-if qui utilise une computed (qui elle effectue une vérification sur le champ).  Dans votre cas le dynamic (ou un composant avis disons) serait affiché ou non en fonction du résultat de la computed qui va être exécutée dès qu'un élément réactif qu'elle contient change.
Tout est dynamique.

**Reply by eparentrq** [reply]
Merci de l'info.

Le problème est que lorsque je sors du champ, le computed n'a pas l'air d'être appelé.
J'en comprends qu'il est appelé au chargement (qui ne m'est pas utile), mais qu'il est aussi appelé lors du changement (ce qui n'arrive pas)
[RDC025.v17.form.txt](https://github.com/MTESSDev/FRW/files/11517960/RDC025.v17.form.txt)

Merci d'y jeter un coup d'oeil.

**Reply by j0ber** [reply]
Quel v-if ne fonctionne pas au juste? et de quelle computed on parle?

**Reply by eparentrq** [reply]
Celui du champ texte pour y mettre l'année.
Le computed fonctionne au chargement, mais pas au changement du champ texte (tous les v-if, sauf celui === 999, je voulais faire un test dans le return, je retournais 9999)

**Reply by j0ber** [reply]
Votre computed ne contient AUCUN élément réactif. C'est normal qu'elle ne soit pas appelée.
Dans votre cas il faudrait prendre le code à l'intérieur de DifferenceAnnee et le mettre dans votre computed AnneeVisee.

À titre d'info, dans votre cas, l'élément réactif est this.form.dateannee.

Votre méthode differenceAnnee ne sert à rien.

**Reply by jodionrq** [reply]
Bonjour,
Est-ce que c'est possible de prendre un champ de type "date" et de changer sont format pour entrer seulement les années (YYYY) ?
Est-ce que le fait d'avoir un champ qui est déjà formaté en années seulement aiderait à l'injection du JS? Exemple :
![image](https://github.com/MTESSDev/FRW/assets/124822406/442f28a8-52ef-4d04-b5d7-52127a439110)

**Reply by j0ber** [reply]
La réponse est NON à vos 2 questions. Ce que vous voulez faire n'est vraiment pas compliqué avec un champ texte qui oblige un format YYYY comme vous l'avez fait.

**Reply by jodionrq** [reply]
Bonjour, j'ai vu dans le P700U qu'il y a un bel ajout de validation personnalisée pour le champ date en format YYYY. Ça semble répondre à nos besoins je vais regarder ça de plus près. Merci!

**Reply by j0ber** [reply]
Bonjour, votre besoin n'était pas une validation, mais l'affichage d'un avertissement il me semble. Ce n'est pas la même chose.

**Reply by jodionrq** [reply]
En fait ça répond à une partie de nos besoins c'est vrai. Au moins on peu afficher un message de validation selon l'année en cours avec le getFullYear.

**Reply by j0ber** [reply]
Comme je vous avait déjà dit afficher un message d'avertissement est encore plus simple dans votre cas. Vous avez tout ce qu'il faut pour le faire et c'est simple.
